pkg/controller/backupinfrastructure/backup_infrastructure_control.go

```diff
@@ -0,0 +1,372 @@
+	)
+	if e := f.Execute(); e != nil {
+		e.Description = fmt.Sprintf("Failed to delete Backup Infrastructure: %s", e.Description)
+		return e
+	}
+
+	o.Logger.Infof("Successfully deleted Backup Infrastructure '%s'", o.BackupInfrastructure.Name)
+	return nil
+}
+
+func (c *defaultControl) updateBackupInfrastructureStatus(o *operation.Operation, state gardenv1beta1.ShootLastOperationState, operationType gardenv1beta1.ShootLastOperationType, operationDescription string, progress int, lastError *gardenv1beta1.LastError) error {
+	if state == gardenv1beta1.ShootLastOperationStateError && o.BackupInfrastructure.Status.LastOperation != nil {
+		progress = o.BackupInfrastructure.Status.LastOperation.Progress
+	}
+	o.BackupInfrastructure.Status.LastOperation = &gardenv1beta1.LastOperation{
+		Type:           operationType,
+		State:          state,
+		Progress:       progress,
+		Description:    operationDescription,
+		LastUpdateTime: metav1.Now(),
+	}
+	o.BackupInfrastructure.Status.LastError = lastError
+	o.BackupInfrastructure.Status.ObservedGeneration = o.BackupInfrastructure.Generation
+
+	newBackupInfrastructure, err := c.updater.UpdateBackupInfrastructureStatus(o.BackupInfrastructure)
+	if err == nil {
+		o.BackupInfrastructure = newBackupInfrastructure
+	}
+	return err
+}
+
+func (c *defaultControl) removeFinalizer(op *operation.Operation) error {
+	// Remove finalizer from BackupInfrastructure
+	backupInfrastructureFinalizers := sets.NewString(op.BackupInfrastructure.Finalizers...)
+	backupInfrastructureFinalizers.Delete(gardenv1beta1.GardenerName)
+	op.BackupInfrastructure.Finalizers = backupInfrastructureFinalizers.UnsortedList()
+
+	newBackupInfrastructure, err := c.k8sGardenClient.GardenClientset().GardenV1beta1().BackupInfrastructures(op.BackupInfrastructure.Namespace).Update(op.BackupInfrastructure)
+	if err != nil {
+		op.Logger.Errorf("Could not remove finalizer of the BackupInfrastructure: %+v", err.Error())
+		return err
+	}
+	op.BackupInfrastructure = newBackupInfrastructure
+	// Wait until the above modifications are reflected in the cache to prevent unwanted reconcile
+	// operations (sometimes the cache is not synced fast enough).
+	return wait.PollImmediate(time.Second, 30*time.Second, func() (bool, error) {
+		backupInfrastructure, err := c.k8sGardenInformers.BackupInfrastructures().Lister().BackupInfrastructures(op.BackupInfrastructure.Namespace).Get(op.BackupInfrastructure.Name)
+		if apierrors.IsNotFound(err) {
+			return true, nil
+		}
+		if err != nil {
+			return false, err
+		}
+		if !sets.NewString(backupInfrastructure.Finalizers...).Has(gardenv1beta1.GardenerName) {
+			return true, nil
+		}
+		return false, nil
+	})
+}
+
+func formatError(message string, err error) *gardenv1beta1.LastError {
+	return &gardenv1beta1.LastError{
+		Description: fmt.Sprintf("%s (%s)", message, err.Error()),
+	}
+}
+
+func computeOperationType(lastOperation *gardenv1beta1.LastOperation) gardenv1beta1.ShootLastOperationType {
```

Comment:
This function has been duplicated (https://github.com/gardener/gardener/blob/master/pkg/controller/shoot/shoot_utils.go#L58-L63). Can we please move it to a general util file? https://github.com/gardener/gardener/tree/master/pkg/controller/utils ?